ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,89 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+class ImpendingTimeout(Exception):
+    pass
+
+
+def crash_flush(_, __):
+    """Raises an `ImpendingTimeout` error which causes the tracer to shutdown."""
+    # TODO: Update to tag error in local root parent instead of raising the error.
+
+    raise ImpendingTimeout("Datadog detected an Impending Timeout")
+
+
+def check_timeout(context):
+    """
+    Creates a timeout to detect when an AWS Lambda handler's remaining
+    time is about to end.
+
+    Crashes flushes when the signal is activated
+    """
+    signal.signal(signal.SIGALRM, crash_flush)
```

Comment:
Only one handler can be set per signal. This operation will overwrite any existing SIGALRM handlers. Also this signal will only work on [unix](https://docs.python.org/3/library/signal.html#signal.SIGALRM) (for example windows will not be supported). We might need to consider a different approach.

Reply:
Instead of using a timer can we register an [exit handler](https://docs.python.org/3/library/atexit.html) that closes open serverless spans. Would this work for you?

Reply:
An approach like [this](https://stackoverflow.com/a/12018594/735204) could fix the issue of this signal handler clobbering preexisting handlers.

Reply:
I instrumented a Lambda function which uses signals and times out. It seems that no user signal is being replaced. Since they're imported during runtime, our signal gets ditched. That said, the only user who would not receive traces in a timeout invocation would be the one meeting the following criteria:
1. registering a `SIGALRM` in a lambda function.
2. function runs out of time.

Is the fix really needed since the only signal being replaced is ours?
The Serverless team is aware of this and agrees on going forward.

Reply:
I think we should follow the best practice of not overriding an existing signal handler. It's not hard for us to call an existing one (following @emmettbutler's approach). 

There is likely some risk since it appears that this signal handler is one of the few ways to run code on timeout in a Lambda function (surely we're not going to be the only ones).

We also have to hope that the user or other libraries give the same courtesy in order for this solution to work.

Reply:
Closing this one based on the demo I gave today explaining that we are not overriding any signals from the users. And other feedback provided around doing this particular approach, feel free to re-open.

Reply:
I think we're not on the same page. My concern with the code as-is, we will overwrite any existing SIGALRM signal handler that is in place by the time we get to it. The signal handler could be added by other libraries that the user installs that execute prior to ours. We can mitigate this by using a solution like https://stackoverflow.com/a/12018594/735204

Reply:
Thanks, I was not understanding what you meant. 
Added a helper method to avoid overriding previous defined `SIGALRM` signals.